### PR TITLE
UCT/IB: Re-factored RC/DC AM short IOV implementation for device memory.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Re-factored RC/DC AM short IOV implementation for device memory.

## Why ?
According to [review comment](https://github.com/openucx/ucx/pull/6279#pullrequestreview-585460141) to reduce code duplication.

## How ?
Combined RC/DC AM short IOV implementation for device memory to single function.
Introduced `uct_mlx5_txqp_bcopy_post` to move common code from RC and DC specific functions.

### CODE REVIEW

File: src/uct/ib/rc/accel/rc_mlx5_ep.c [hunk 1]
```diff
@@ -47,12 +45,9 @@ uct_rc_mlx5_txqp_bcopy_post(uct_rc_mlx5_iface_common_t *iface,
                             uct_rc_iface_send_desc_t *desc, const void *buffer,
                             uct_ib_log_sge_t *log_sge)
 {
-    desc->super.sn = txwq->sw_pi;
-    uct_rc_mlx5_txqp_dptr_post(iface, IBV_QPT_RC, txqp, txwq,
-                               opcode, buffer, length, &desc->lkey,
-                               rdma_raddr, rdma_rkey, 0, 0, 0, 0,
-                               NULL, NULL, 0, fm_ce_se, imm_val_be, INT_MAX, log_sge);
-    uct_rc_txqp_add_send_op(txqp, &desc->super);
+    uct_mlx5_txqp_bcopy_post(iface, IBV_QPT_RC, txqp, txwq, opcode, length,
```
Comment (brminich) [inline src/uct/ib/rc/accel/rc_mlx5_ep.c:48]: no need for `uct_rc_mlx5_txqp_bcopy_post` now, can use `uct_mlx5_txqp_bcopy_post` directly

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 2]
```diff
@@ -28,15 +28,12 @@ uct_dc_mlx5_iface_bcopy_post(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep,
...
-    desc->super.sn = txwq->sw_pi;
-    uct_rc_mlx5_txqp_dptr_post(&iface->super, UCT_IB_QPT_DCI, txqp, txwq,
-                               opcode, buffer, length, &desc->lkey,
-                               rdma_raddr, rdma_rkey, 0, 0, 0, 0,
-                               &ep->av, uct_dc_mlx5_ep_get_grh(ep),
-                               uct_ib_mlx5_wqe_av_size(&ep->av),
-                               MLX5_WQE_CTRL_CQ_UPDATE | send_flags, imm_val_be, INT_MAX,
-                               log_sge);
-    uct_rc_txqp_add_send_op(txqp, &desc->super);
+    uct_mlx5_txqp_bcopy_post(&iface->super, UCT_IB_QPT_DCI, txqp, txwq, opcode,
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.c:31]: do we really need `uct_dc_mlx5_iface_bcopy_post` wrapper now? Seems can just use `uct_mlx5_txqp_bcopy_post` and declare `txqp, txwq` in place

**Review by brminich — APPROVED** [review]

**Comment by rakhmets** [comment]
bot:pipe:retest

**Comment by rakhmets** [comment]
@yosefe could you please to review the changes.

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 3]
```diff
@@ -303,38 +281,13 @@ uct_dc_mlx5_ep_short_dm(uct_dc_mlx5_ep_t *ep, uct_rc_mlx5_dm_copy_data_t *cache,
...
-                                             &buffer, &log_sge);
-    if (ucs_unlikely(UCS_STATUS_IS_ERR(status))) {
-        return status;
-    }
-
-    uct_dc_mlx5_iface_bcopy_post(
-            iface, ep, MLX5_OPCODE_SEND, hdr_len + iov_length, 0, 0, desc,
-            MLX5_WQE_CTRL_SOLICITED | MLX5_WQE_CTRL_CQ_UPDATE, 0, buffer,
-            (log_sge.num_sge > 0) ? &log_sge : NULL);
+    UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:284]: can we also unite the "uct_dc_mlx5_ep_short_dm" function ?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 4]
```diff
@@ -600,9 +561,12 @@ ucs_status_t uct_dc_mlx5_ep_get_bcopy(uct_ep_h tl_ep,
     UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
     uct_rc_mlx5_ep_fence_get(&iface->super, txwq, &rkey, &fm_ce_se);
 
-    uct_dc_mlx5_iface_bcopy_post(iface, ep, MLX5_OPCODE_RDMA_READ, length,
-                                 remote_addr, rkey, desc, fm_ce_se, 0,
-                                 desc + 1, NULL);
+    uct_mlx5_txqp_bcopy_post(&iface->super, UCT_IB_QPT_DCI, txqp, txwq,
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:564]: i think we can still keep uct_dc_mlx5_iface_bcopy_post, but just as wrapper over uct_mlx5_txqp_bcopy_post

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 5]
```diff
@@ -1715,6 +1735,44 @@ static ucs_status_t UCS_F_ALWAYS_INLINE uct_rc_mlx5_common_dm_make_data(
     *buffer_p = buffer;
     return UCS_OK;
 }
+
+static ucs_status_t UCS_F_ALWAYS_INLINE uct_mlx5_ep_am_short_iov_dm(
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl:1739]: pls rename to uct_rc_mlx5_common_ep_am_short_iov_dm

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 6]
```diff
@@ -909,6 +909,26 @@ void uct_rc_mlx5_txqp_dptr_post_iov(uct_rc_mlx5_iface_common_t *iface, int qp_ty
                                  max_log_sge, NULL);
 }
 
+/*
+ * Helper function for buffer-copy post.
+ * Adds the descriptor to the callback queue.
+ */
+static UCS_F_ALWAYS_INLINE void uct_mlx5_txqp_bcopy_post(
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl:916]: pls rename to uct_rc_mlx5_common_txqp_bcopy_post

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 3 above)
Comment (rakhmets) [reply src/uct/ib/dc/dc_mlx5_ep.c:284]: With `uct_rc_mlx5_ep_short_dm`?

Comment (rakhmets) [reply src/uct/ib/dc/dc_mlx5_ep.c]: @yosefe I've combined `uct_dc_mlx5_ep_short_dm` and `uct_rc_mlx5_ep_short_dm` to `uct_rc_mlx5_common_ep_short_dm`

**Review by yosefe — APPROVED** [review]
nice!
pls squash

**Comment by rakhmets** [comment]
bot:pipe:retest

**Comment by rakhmets** [comment]
@yosefe this PR is ready to be merged.

**Pull request merged**